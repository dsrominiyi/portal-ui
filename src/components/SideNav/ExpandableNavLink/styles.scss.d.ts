export const expandableNavLink: string;
export const dropDown: string;
export const mini: string;
export const growIn: string;
export const expanded: string;
export const hidden: string;
export const link: string;
export const clickableLink: string;
export const chevron: string;
