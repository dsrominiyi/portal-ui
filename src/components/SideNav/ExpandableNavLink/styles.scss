@import '../../../style/vars';

.expandable-nav-link {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.drop-down {
  width: calc(100% - 2rem);
  border-radius: 0.35rem;
  background-color: $brand-white;
  overflow: hidden;
  font-size: 0.85rem;
  transition: height 0.3s, margin 0.3s;

  > :first-child {
    margin-top: 0.5rem;
  }

  > :last-child {
    margin-bottom: 0.5rem;
  }

  &.mini {
    position: absolute;
    left: calc(104px + 0.75rem);
    width: auto;
    overflow: hidden;
    white-space: nowrap;
    box-shadow: 0 0.15rem 1.75rem 0 rgba(58, 59, 69, 0.15);
    transition: none;
    animation-name: growIn;
    animation-duration: 0.5s;
    z-index: 1;
  }

  &.expanded {
    margin-bottom: 1rem;
  }

  &.hidden {
    display: none;
  }
}

.link {
  display: block;
  color: $brand-heavy-thunder;
  background-color: $brand-white;
  border-radius: 0.35rem;
  margin: 0 0.5rem;
  padding: 0.5rem 1rem;
  font-size: 0.85rem;
  text-decoration: none;
  cursor: pointer;

  &:hover {
    color: $brand-heavy-thunder;
    background-color: $brand-grey-white;
  }
}

.clickable-link {
  position: relative;
  width: 100%;
  display: flex;
  align-items: center;
  color: $brand-white;

  .chevron {
    position: absolute;
    right: 1rem;
    pointer-events: none;
    color: #b7b9cc;
    transition: transform 0.3s;

    &.expanded {
      transform: rotate(90deg);
    }
  }

  > * {
    flex: 1;
  }
}

@keyframes growIn {
  0% {
    transform: scale(0.9);
    opacity: 0;
  }

  100% {
    transform: scale(1);
    opacity: 1;
  }
}
