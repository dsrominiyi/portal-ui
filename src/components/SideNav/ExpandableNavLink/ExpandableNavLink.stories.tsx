import React, { ReactNode } from 'react';
import { storiesOf } from '@storybook/react';
import ExpandableNavLink, { SubItem } from '.';
import styles from '../../../../.storybook/styles.scss';

const stories = storiesOf('ExpandableNavLink', module);

export const render = (
  label: string,
  icon: ReactNode,
  subItems: SubItem[],
  isMiniLink?: boolean
) => () => (
  <ExpandableNavLink label={label} icon={icon} subItems={subItems} isMiniLink={isMiniLink} />
);

const renderWithBackground = (renderFunc: () => ReactNode, mini?: boolean) => () => (
  <div className={styles.darkBackground} style={{ width: mini ? 104 : 224 }}>
    {renderFunc()}
  </div>
);

export const testItems = [
  { label: 'New Individual', href: '/newind', onClick: () => alert('You Clicked') },
  { label: 'New Company', href: '/newcompany', onClick: () => alert('You Clicked') },
  { label: 'Find Customer', href: '/search', onClick: () => alert('You Clicked') }
];

const withoutHrefItems = [
  { label: 'New Individual', onClick: () => alert('You Clicked') },
  { label: 'New Company', onClick: () => alert('You Clicked') },
  { label: 'Find Customer', onClick: () => alert('You Clicked') }
];

export const storyRenders = {
  defaultSize: render('Customer', <i className="fas fa-user" />, testItems),
  miniLink: render('Customer', <i className="fas fa-user" />, testItems, true),
  withoutHrefs: render('Customer', <i className="fas fa-user" />, withoutHrefItems)
};

stories.add('Default size', renderWithBackground(storyRenders.defaultSize));
stories.add('Mini link', renderWithBackground(storyRenders.miniLink, true));
stories.add('Without Hrefs', renderWithBackground(storyRenders.withoutHrefs));
