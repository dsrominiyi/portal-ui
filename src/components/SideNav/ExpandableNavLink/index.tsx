import React, {
  FunctionComponent,
  ReactNode,
  useState,
  useRef,
  useEffect,
  MouseEvent
} from 'react';
import classnames from 'classnames/bind';
import Link from '../../Link';
import NavLink from '../NavLink';
import styles from './styles.scss';

const cx = classnames.bind(styles);

export interface SubItem {
  label: string;
  href?: string;
  onClick?(event: MouseEvent): void;
}

export interface Props {
  label: string;
  icon: ReactNode;
  isActive?: boolean;
  isMiniLink?: boolean;
  subItems: SubItem[];
}

export const ExpandableNavLink: FunctionComponent<Props> = ({
  label,
  icon,
  isActive,
  isMiniLink,
  subItems
}: Props) => {
  const [isExpanded, toggleExpand] = useState(false);
  const dropdownRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    const dropdownDiv = dropdownRef.current as HTMLInputElement;
    if (!isMiniLink) {
      dropdownDiv.style.height = isExpanded ? `${dropdownDiv.scrollHeight}px` : '0';
    } else {
      dropdownDiv.style.height = 'auto';
    }
  }, [isExpanded, isMiniLink]);

  useEffect(() => {
    if (isExpanded) {
      toggleExpand(!isExpanded);
    }
  }, [isMiniLink]);

  const displayLinks = subItems.map((subItem, index) => {
    return (
      <Link
        key={index}
        label={subItem.label}
        href={subItem.href}
        onClick={subItem.href ? undefined : subItem.onClick}
        classNames={[styles.link]}
      />
    );
  });

  return (
    <div className={styles.expandableNavLink}>
      <div
        className={cx({ clickableLink: true, mini: isMiniLink })}
        onClick={() => toggleExpand(!isExpanded)}
      >
        <NavLink label={label} icon={icon} isActive={isActive} isMiniLink={isMiniLink} />
        {!isMiniLink && (
          <i
            className={cx({ chevron: true, 'fas fa-chevron-right': true, expanded: isExpanded })}
          />
        )}
      </div>
      <div
        ref={dropdownRef}
        className={cx({
          dropDown: true,
          mini: isMiniLink,
          expanded: !isMiniLink && isExpanded,
          hidden: isMiniLink && !isExpanded
        })}
      >
        {displayLinks}
      </div>
    </div>
  );
};

ExpandableNavLink.displayName = 'ExpandableNavLink';
export default ExpandableNavLink;
