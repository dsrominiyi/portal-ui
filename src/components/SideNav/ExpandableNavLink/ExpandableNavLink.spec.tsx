import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { storyRenders, render, testItems } from './ExpandableNavLink.stories';
import testStorySnapshots from '../../../helpers/test/testStorySnapshots';

let component: ReactWrapper;

describe('ExpandableNavLink', () => {
  test('component story renders match snapshots', () => {
    testStorySnapshots(storyRenders);
  });

  describe('ExpandableNavLink functionality', () => {
    beforeEach(() => {
      component = mount(render('Customer', <i className="fas fa-user" />, testItems)());
    });

    test('clicking the NavLink adds the expanded class to the chevron icon', () => {
      component.find('.clickableLink').simulate('click');
      const chevronIcon = component.find('.chevron');

      expect(chevronIcon.props().className).toContain('expanded');
    });

    test('clicking the NavLink expands the dropdown', () => {
      const scrollHeight = 105;
      let dropDown = component.find('.dropDown');
      Object.defineProperty(dropDown.getDOMNode(), 'scrollHeight', { value: scrollHeight });

      component.find('.clickableLink').simulate('click');
      dropDown = component.find('.dropDown');

      expect(dropDown.getDOMNode().getAttribute('style')).toBe(`height: ${scrollHeight}px;`);
    });

    test('clicking the expanded NavLink retracts the dropdown', () => {
      let dropDown = component.find('.dropDown');
      Object.defineProperty(dropDown.getDOMNode(), 'scrollHeight', { value: 105 });

      const clickableLink = component.find('.clickableLink');
      clickableLink.simulate('click');
      clickableLink.simulate('click');
      dropDown = component.find('.dropDown');

      expect(dropDown.getDOMNode().getAttribute('style')).toBe('height: 0px;');
    });

    test('clicking the mini link sets the dropdown height property to auto', () => {
      component = mount(render('Customer', <i className="fas fa-user" />, testItems, true)());
      component.find('.clickableLink').simulate('click');
      const dropDown = component.find('.dropDown');

      expect(dropDown.getDOMNode().getAttribute('style')).toBe('height: auto;');
    });

    test('minimising the mini link adds the hidden class to dropdown', () => {
      component = mount(render('Customer', <i className="fas fa-user" />, testItems, true)());
      const clickableLink = component.find('.clickableLink');
      clickableLink.simulate('click');
      clickableLink.simulate('click');
      const dropDown = component.find('.dropDown');

      expect(dropDown.props().className).toContain('hidden');
    });

    test('changing the isMiniLink prop retracts/minimises the dropdown', () => {
      const clickableLink = component.find('.clickableLink');
      clickableLink.simulate('click');

      component.setProps({ isMiniLink: true });
      component.update();
      let dropDown = component.find('.dropDown');

      expect(dropDown.props().className).toContain('hidden');

      clickableLink.simulate('click');
      component.setProps({ isMiniLink: false });
      component.update();
      dropDown = component.find('.dropDown');

      expect(dropDown.getDOMNode().getAttribute('style')).toBe('height: 0px;');
    });
  });
});
