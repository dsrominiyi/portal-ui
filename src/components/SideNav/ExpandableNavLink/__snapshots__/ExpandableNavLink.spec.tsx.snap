// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandableNavLink component story renders match snapshots: defaultSize 1`] = `
<div
  className="expandableNavLink"
>
  <div
    className="clickableLink"
    onClick={[Function]}
  >
    <NavLink
      icon={
        <i
          className="fas fa-user"
        />
      }
      label="Customer"
    />
    <i
      className="chevron fas fa-chevron-right"
    />
  </div>
  <div
    className="dropDown"
  >
    <Link
      classNames={
        Array [
          "link",
        ]
      }
      href="/newind"
      key="0"
      label="New Individual"
    />
    <Link
      classNames={
        Array [
          "link",
        ]
      }
      href="/newcompany"
      key="1"
      label="New Company"
    />
    <Link
      classNames={
        Array [
          "link",
        ]
      }
      href="/search"
      key="2"
      label="Find Customer"
    />
  </div>
</div>
`;

exports[`ExpandableNavLink component story renders match snapshots: miniLink 1`] = `
<div
  className="expandableNavLink"
>
  <div
    className="clickableLink mini"
    onClick={[Function]}
  >
    <NavLink
      icon={
        <i
          className="fas fa-user"
        />
      }
      isMiniLink={true}
      label="Customer"
    />
  </div>
  <div
    className="dropDown mini hidden"
  >
    <Link
      classNames={
        Array [
          "link",
        ]
      }
      href="/newind"
      key="0"
      label="New Individual"
    />
    <Link
      classNames={
        Array [
          "link",
        ]
      }
      href="/newcompany"
      key="1"
      label="New Company"
    />
    <Link
      classNames={
        Array [
          "link",
        ]
      }
      href="/search"
      key="2"
      label="Find Customer"
    />
  </div>
</div>
`;

exports[`ExpandableNavLink component story renders match snapshots: withoutHrefs 1`] = `
<div
  className="expandableNavLink"
>
  <div
    className="clickableLink"
    onClick={[Function]}
  >
    <NavLink
      icon={
        <i
          className="fas fa-user"
        />
      }
      label="Customer"
    />
    <i
      className="chevron fas fa-chevron-right"
    />
  </div>
  <div
    className="dropDown"
  >
    <Link
      classNames={
        Array [
          "link",
        ]
      }
      key="0"
      label="New Individual"
      onClick={[Function]}
    />
    <Link
      classNames={
        Array [
          "link",
        ]
      }
      key="1"
      label="New Company"
      onClick={[Function]}
    />
    <Link
      classNames={
        Array [
          "link",
        ]
      }
      key="2"
      label="Find Customer"
      onClick={[Function]}
    />
  </div>
</div>
`;
